# Download and prepare the data. Explain the procedure.

### Project description

To produce gold a row ore goes in 3 steps where the second and third are part of step calls purification and the first calls flotation. In every step a raw material enters and two streams leave. One is with the concentrate and this is the main process line and where most of the gold stays and the second calls tails and there are all the leftover that will leave the process and if any gold passed there this is a loss and it will cause a reduction in the recovery.

![flowchart](img/flowchart.jpg)

### Import needed libraries

In [2]:
# Data tools
import pandas as pd
import numpy as np

# Graphics and display
from IPython.core.interactiveshell import InteractiveShell
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.figure_factory as ff
%matplotlib inline

# Ml
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.dummy import DummyRegressor

# Statistics
from scipy import stats

print('Project libraries has been successfully been imported!')

Project libraries has been successfully been imported!


 ### Set environment variables¶

In [3]:
# set to display all output not only print() or last output
InteractiveShell.ast_node_interactivity = "all"        

## Open the files and look into the data.

In [4]:
try:
    data_full = pd.read_csv('gold_recovery_full.csv')
    data_test = pd.read_csv('gold_recovery_test.csv')
    data_train = pd.read_csv('gold_recovery_train.csv')
except:
    data_full = pd.read_csv('/datasets/gold_recovery_full.csv')
    data_test = pd.read_csv('/datasets/gold_recovery_test.csv')
    data_train = pd.read_csv('/datasets/gold_recovery_train.csv')
    
print('Data has been read correctly!')

Data has been read correctly!


### Looking at the dataset

### Headers of the data

There are many features in this project and to easily tells what is in each one there is a formal naming system:

[stage].[parameter_type].[parameter_name]

Example: rougher.input.feed_ag


### Data describe

In [5]:
# Checking 0's
def zero_check(df):
    for i in df:
        print(i,len(df[df[i]==0]))
        
# function to determine if columns in file have null values        
def get_percent_of_na(df, num):
    count = 0
    df = df.copy()
    s = (df.isna().sum() / df.shape[0])
    for column, percent in zip(s.index, s.values):
        num_of_nulls = df[column].isna().sum()
        if num_of_nulls == 0:
            continue
        else:
            count += 1
        print('{} has {} nulls, which is {:.{}%} percent of Nulls'.format(column, num_of_nulls, percent, num))
    if count != 0:
        print("\033[1m" + 'There are {} columns with NA.'.format(count) + "\033[0m")
    else:
        print()
        print("\033[1m" + 'There are no columns with NA.' + "\033[0m")       
        
# function to display general information about the dataset
def general_info(df):
    print("\033[1m" + "\033[0m")
    display(pd.concat([df.dtypes, df.count(),df.isna().sum(),df.isna().sum()/len(df)], keys=['type','count','na','na%'],
                      axis=1))
    print()
    print("\033[1m" + 'Head:')  
    display(df.head())
    print()
    print("\033[1m" + 'Tail:')
    display(df.tail())
    print()
    print("\033[1m" + 'Info:')
    print()
    display(df.info())
    print()
    print("\033[1m" + 'Describe:')
    print()
    display(df.describe())
    print()
    print("\033[1m" + 'Describe include: all :')
    print()
    display(df.describe(include='all'))
    print()
    print("\033[1m" + 'nulls in the columns:')
    print()
    display(get_percent_of_na(df, 4))  # check this out
    print()
    print("\033[1m" + 'Zeros in the columns:') 
    print()
    display(zero_check(df))
    print()
    print("\033[1m" + 'Shape:', df.shape)
    print()
    print()
    print('Duplicated:',"\033[1m" + 'We have {} duplicated rows\n'.format(df.duplicated().sum()) + "\033[0m")
    print()
    print("\033[1m" + 'Dtypes:')  
    display(df.dtypes)
    print()

#### data_full

In [6]:
#print our info data
print('information about "data_full" dataset:')
general_info(data_full)

information about "data_full" dataset:



type  count   na       na%
date                                         object  22716    0  0.000000
final.output.concentrate_ag                 float64  22627   89  0.003918
final.output.concentrate_pb                 float64  22629   87  0.003830
final.output.concentrate_sol                float64  22331  385  0.016948
final.output.concentrate_au                 float64  22630   86  0.003786
...                                             ...    ...  ...       ...
secondary_cleaner.state.floatbank5_a_level  float64  22615  101  0.004446
secondary_cleaner.state.floatbank5_b_air    float64  22615  101  0.004446
secondary_cleaner.state.floatbank5_b_level  float64  22616  100  0.004402
secondary_cleaner.state.floatbank6_a_air    float64  22597  119  0.005239
secondary_cleaner.state.floatbank6_a_level  float64  22615  101  0.004446

[87 rows x 4 columns]


Head:


date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    


Tail:


date  final.output.concentrate_ag  \
22711  2018-08-18 06:59:59                     3.224920   
22712  2018-08-18 07:59:59                     3.195978   
22713  2018-08-18 08:59:59                     3.109998   
22714  2018-08-18 09:59:59                     3.367241   
22715  2018-08-18 10:59:59                     3.598375   

       final.output.concentrate_pb  final.output.concentrate_sol  \
22711                    11.356233                      6.803482   
22712                    11.349355                      6.862249   
22713                    11.434366                      6.886013   
22714                    11.625587                      6.799433   
22715                    11.737832                      6.717509   

       final.output.concentrate_au  final.output.recovery  \
22711                    46.713954              73.755150   
22712                    46.866780              69.049291   
22713                    46.795691              67.002189   
22714                    46.408188              65.523246   
22715                    46.299438              70.281454   

       final.output.tail_ag  final.output.tail_pb  final.output.tail_sol  \
22711              8.769645              3.141541              10.403181   
22712              8.897321              3.130493              10.549470   
22713              8.529606              2.911418              11.115147   
22714              8.777171              2.819214              10.463847   
22715              8.406690              2.517518              10.652193   

       final.output.tail_au  ...  secondary_cleaner.state.floatbank4_a_air  \
22711              1.529220  ...                                 23.031497   
22712              1.612542  ...                                 22.960095   
22713              1.596616  ...                                 23.015718   
22714              1.602879  ...                                 23.024963   
22715              1.389434  ...                                 23.018622   

       secondary_cleaner.state.floatbank4_a_level  \
22711                                 -501.167942   
22712                                 -501.612783   
22713                                 -501.711599   
22714                                 -501.153409   
22715                                 -500.492702   

       secondary_cleaner.state.floatbank4_b_air  \
22711                                 20.007571   
22712                                 20.035660   
22713                                 19.951231   
22714                                 20.054122   
22715                                 20.020205   

       secondary_cleaner.state.floatbank4_b_level  \
22711                                 -499.740028   
22712                                 -500.251357   
22713                                 -499.857027   
22714                                 -500.314711   
22715                                 -500.220296   

       secondary_cleaner.state.floatbank5_a_air  \
22711                                 18.006038   
22712                                 17.998535   
22713                                 18.019543   
22714                                 17.979515   
22715                                 17.963512   

       secondary_cleaner.state.floatbank5_a_level  \
22711                                 -499.834374   
22712                                 -500.395178   
22713                                 -500.451156   
22714                                 -499.272871   
22715                                 -499.939490   

       secondary_cleaner.state.floatbank5_b_air  \
22711                                 13.001114   
22712                                 12.954048   
22713                                 13.023431   
22714                                 12.992404   
22715                                 12.990306   

       secondary_cleaner.state.floatbank5_b_level  \
22711                             


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22716 entries, 0 to 22715
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                22716 non-null  object 
 1   final.output.concentrate_ag                         22627 non-null  float64
 2   final.output.concentrate_pb                         22629 non-null  float64
 3   final.output.concentrate_sol                        22331 non-null  float64
 4   final.output.concentrate_au                         22630 non-null  float64
 5   final.output.recovery                               20753 non-null  float64
 6   final.output.tail_ag                                22633 non-null  float64
 7   final.output.tail_pb                                22516 non-null  float64
 8   final.output.tail_sol                               22445 non-null  

None


Describe:



final.output.concentrate_ag  final.output.concentrate_pb  \
count                 22627.000000                 22629.000000   
mean                      4.781559                     9.095308   
std                       2.030128                     3.230797   
min                       0.000000                     0.000000   
25%                       4.018525                     8.750171   
50%                       4.953729                     9.914519   
75%                       5.862593                    10.929839   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  22331.000000                 22630.000000   
mean                       8.640317                    40.001172   
std                        3.785035                    13.398062   
min                        0.000000                     0.000000   
25%                        7.116799                    42.383721   
50%                        8.908792                    44.653436   
75%                       10.705824                    46.111999   
max                       19.615720                    53.611374   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           20753.000000          22633.000000          22516.000000   
mean               67.447488              8.923690              2.488252   
std                11.616034              3.517917              1.189407   
min                 0.000000              0.000000              0.000000   
25%                63.282393              7.684016              1.805376   
50%                68.322258              9.484369              2.653001   
75%                72.950836             11.084557              3.287790   
max               100.000000             19.552149              6.086532   

       final.output.tail_sol  final.output.tail_au  \
count           22445.000000          22635.000000   
mean                9.523632              2.827459   
std                 4.079739              1.262834   
min                 0.000000              0.000000   
25%                 8.143576              2.303108   
50%                10.212998              2.913794   
75%                11.860824              3.555077   
max                22.861749              9.789625   

       primary_cleaner.input.sulfate  ...  \
count                   21107.000000  ...   
mean                      140.277672  ...   
std                        49.919004  ...   
min                         0.000003  ...   
25%                       110.177081  ...   
50%                       141.330501  ...   
75%                       174.049914  ...   
max                       274.409626  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              22571.000000   
mean                                  18.205125   
std                                    6.560700   
min                                    0.000000   
25%                                   14.095940   
50%                                   18.007326   
75%                                   22.998194   
max                                   60.000000   

       secondary_cleaner.state.floatbank4_a_level  \
count                                22587.000000   
mean                                  -499.878977   
std                                     80.273964   
min                                   -799.920713   
25%                                   -500.896232   
50%                                   -499.917108   
75%                                   -498.361545   
max                                   -127.692333   

       secondary_cleaner.state.floatbank4_b_air  \
count                              22608.000000   
mean                                  14.356474   
std                                    5.655791   
min                                    0.000000   
25%                


Describe include: all :



date  final.output.concentrate_ag  \
count                 22716                 22627.000000   
unique                22716                          NaN   
top     2016-01-15 00:00:00                          NaN   
freq                      1                          NaN   
mean                    NaN                     4.781559   
std                     NaN                     2.030128   
min                     NaN                     0.000000   
25%                     NaN                     4.018525   
50%                     NaN                     4.953729   
75%                     NaN                     5.862593   
max                     NaN                    16.001945   

        final.output.concentrate_pb  final.output.concentrate_sol  \
count                  22629.000000                  22331.000000   
unique                          NaN                           NaN   
top                             NaN                           NaN   
freq                            NaN                           NaN   
mean                       9.095308                      8.640317   
std                        3.230797                      3.785035   
min                        0.000000                      0.000000   
25%                        8.750171                      7.116799   
50%                        9.914519                      8.908792   
75%                       10.929839                     10.705824   
max                       17.031899                     19.615720   

        final.output.concentrate_au  final.output.recovery  \
count                  22630.000000           20753.000000   
unique                          NaN                    NaN   
top                             NaN                    NaN   
freq                            NaN                    NaN   
mean                      40.001172              67.447488   
std                       13.398062              11.616034   
min                        0.000000               0.000000   
25%                       42.383721              63.282393   
50%                       44.653436              68.322258   
75%                       46.111999              72.950836   
max                       53.611374             100.000000   

        final.output.tail_ag  final.output.tail_pb  final.output.tail_sol  \
count           22633.000000          22516.000000           22445.000000   
unique                   NaN                   NaN                    NaN   
top                      NaN                   NaN                    NaN   
freq                     NaN                   NaN                    NaN   
mean                8.923690              2.488252               9.523632   
std                 3.517917              1.189407               4.079739   
min                 0.000000              0.000000               0.000000   
25%                 7.684016              1.805376               8.143576   
50%                 9.484369              2.653001              10.212998   
75%                11.084557              3.287790              11.860824   
max                19.552149              6.086532              22.861749   

        final.output.tail_au  ...  secondary_cleaner.state.floatbank4_a_air  \
count           22635.000000  ...                              22571.000000   
unique                   NaN  ...                                       NaN   
top                      NaN  ...                                       NaN   
freq                     NaN  ...                                       NaN   
mean                2.827459  ...                                 18.205125   
std                 1.262834  ...                                  6.560700   
min                 0.000000  ...                                  0.000000   
25%                 2.303108  ...                                 14.095940   
50%                 2.913794  ...                                 18.007326   
75%          


nulls in the columns:

final.output.concentrate_ag has 89 nulls, which is 0.3918% percent of Nulls
final.output.concentrate_pb has 87 nulls, which is 0.3830% percent of Nulls
final.output.concentrate_sol has 385 nulls, which is 1.6948% percent of Nulls
final.output.concentrate_au has 86 nulls, which is 0.3786% percent of Nulls
final.output.recovery has 1963 nulls, which is 8.6415% percent of Nulls
final.output.tail_ag has 83 nulls, which is 0.3654% percent of Nulls
final.output.tail_pb has 200 nulls, which is 0.8804% percent of Nulls
final.output.tail_sol has 271 nulls, which is 1.1930% percent of Nulls
final.output.tail_au has 81 nulls, which is 0.3566% percent of Nulls
primary_cleaner.input.sulfate has 1609 nulls, which is 7.0831% percent of Nulls
primary_cleaner.input.depressant has 1546 nulls, which is 6.8058% percent of Nulls
primary_cleaner.input.xanthate has 1151 nulls, which is 5.0669% percent of Nulls
primary_cleaner.output.concentrate_ag has 98 nulls, which is 0.4314% percen

None


Zeros in the columns:

date 0
final.output.concentrate_ag 1613
final.output.concentrate_pb 1613
final.output.concentrate_sol 1613
final.output.concentrate_au 1613
final.output.recovery 151
final.output.tail_ag 1950
final.output.tail_pb 1950
final.output.tail_sol 1950
final.output.tail_au 1950
primary_cleaner.input.sulfate 0
primary_cleaner.input.depressant 57
primary_cleaner.input.feed_size 0
primary_cleaner.input.xanthate 0
primary_cleaner.output.concentrate_ag 1626
primary_cleaner.output.concentrate_pb 1626
primary_cleaner.output.concentrate_sol 1626
primary_cleaner.output.concentrate_au 1626
primary_cleaner.output.tail_ag 1953
primary_cleaner.output.tail_pb 1953
primary_cleaner.output.tail_sol 1953
primary_cleaner.output.tail_au 1953
primary_cleaner.state.floatbank8_a_air 349
primary_cleaner.state.floatbank8_a_level 0
primary_cleaner.state.floatbank8_b_air 345
primary_cleaner.state.floatbank8_b_level 0
primary_cleaner.state.floatbank8_c_air 330
primary_cleaner.state.floatbank8_c_le

None


Shape: (22716, 87)


Duplicated: We have 0 duplicated rows


Dtypes:


date                                           object
final.output.concentrate_ag                   float64
final.output.concentrate_pb                   float64
final.output.concentrate_sol                  float64
final.output.concentrate_au                   float64
                                               ...   
secondary_cleaner.state.floatbank5_a_level    float64
secondary_cleaner.state.floatbank5_b_air      float64
secondary_cleaner.state.floatbank5_b_level    float64
secondary_cleaner.state.floatbank6_a_air      float64
secondary_cleaner.state.floatbank6_a_level    float64
Length: 87, dtype: object

notes
- There are 87 columns in the data
- All values are floats
- Some values are negative. They are reading of levels indicators. Since it's relative information it make sense
- There are 22716 observations in the data
- Almost all columns contains Non values
- Part of the columns contains some zeroes
- No duplicates rows


data_train['new_rougher_recovery'] = (
    data_train['rougher.output.concentrate_au'] * (
        data_train['rougher.input.feed_au'] - data_train['rougher.output.tail_au']) ) \
/ (data_train['rougher.input.feed_au'] * (
    data_train['rougher.output.concentrate_au'] - data_train['rougher.output.tail_au']) )*100

#### data_train

In [7]:
#print our info data
print('information about "data_train" dataset:')
general_info(data_train)

information about "data_train" dataset:



type  count   na       na%
date                                         object  16860    0  0.000000
final.output.concentrate_ag                 float64  16788   72  0.004270
final.output.concentrate_pb                 float64  16788   72  0.004270
final.output.concentrate_sol                float64  16490  370  0.021945
final.output.concentrate_au                 float64  16789   71  0.004211
...                                             ...    ...  ...       ...
secondary_cleaner.state.floatbank5_a_level  float64  16775   85  0.005042
secondary_cleaner.state.floatbank5_b_air    float64  16775   85  0.005042
secondary_cleaner.state.floatbank5_b_level  float64  16776   84  0.004982
secondary_cleaner.state.floatbank6_a_air    float64  16757  103  0.006109
secondary_cleaner.state.floatbank6_a_level  float64  16775   85  0.005042

[87 rows x 4 columns]


Head:


date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    


Tail:


date  final.output.concentrate_ag  \
16855  2018-08-18 06:59:59                     3.224920   
16856  2018-08-18 07:59:59                     3.195978   
16857  2018-08-18 08:59:59                     3.109998   
16858  2018-08-18 09:59:59                     3.367241   
16859  2018-08-18 10:59:59                     3.598375   

       final.output.concentrate_pb  final.output.concentrate_sol  \
16855                    11.356233                      6.803482   
16856                    11.349355                      6.862249   
16857                    11.434366                      6.886013   
16858                    11.625587                      6.799433   
16859                    11.737832                      6.717509   

       final.output.concentrate_au  final.output.recovery  \
16855                    46.713954              73.755150   
16856                    46.866780              69.049291   
16857                    46.795691              67.002189   
16858                    46.408188              65.523246   
16859                    46.299438              70.281454   

       final.output.tail_ag  final.output.tail_pb  final.output.tail_sol  \
16855              8.769645              3.141541              10.403181   
16856              8.897321              3.130493              10.549470   
16857              8.529606              2.911418              11.115147   
16858              8.777171              2.819214              10.463847   
16859              8.406690              2.517518              10.652193   

       final.output.tail_au  ...  secondary_cleaner.state.floatbank4_a_air  \
16855              1.529220  ...                                 23.031497   
16856              1.612542  ...                                 22.960095   
16857              1.596616  ...                                 23.015718   
16858              1.602879  ...                                 23.024963   
16859              1.389434  ...                                 23.018622   

       secondary_cleaner.state.floatbank4_a_level  \
16855                                 -501.167942   
16856                                 -501.612783   
16857                                 -501.711599   
16858                                 -501.153409   
16859                                 -500.492702   

       secondary_cleaner.state.floatbank4_b_air  \
16855                                 20.007571   
16856                                 20.035660   
16857                                 19.951231   
16858                                 20.054122   
16859                                 20.020205   

       secondary_cleaner.state.floatbank4_b_level  \
16855                                 -499.740028   
16856                                 -500.251357   
16857                                 -499.857027   
16858                                 -500.314711   
16859                                 -500.220296   

       secondary_cleaner.state.floatbank5_a_air  \
16855                                 18.006038   
16856                                 17.998535   
16857                                 18.019543   
16858                                 17.979515   
16859                                 17.963512   

       secondary_cleaner.state.floatbank5_a_level  \
16855                                 -499.834374   
16856                                 -500.395178   
16857                                 -500.451156   
16858                                 -499.272871   
16859                                 -499.939490   

       secondary_cleaner.state.floatbank5_b_air  \
16855                                 13.001114   
16856                                 12.954048   
16857                                 13.023431   
16858                                 12.992404   
16859                                 12.990306   

       secondary_cleaner.state.floatbank5_b_level  \
16855                             


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16860 entries, 0 to 16859
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                16860 non-null  object 
 1   final.output.concentrate_ag                         16788 non-null  float64
 2   final.output.concentrate_pb                         16788 non-null  float64
 3   final.output.concentrate_sol                        16490 non-null  float64
 4   final.output.concentrate_au                         16789 non-null  float64
 5   final.output.recovery                               15339 non-null  float64
 6   final.output.tail_ag                                16794 non-null  float64
 7   final.output.tail_pb                                16677 non-null  float64
 8   final.output.tail_sol                               16715 non-null  

None


Describe:



final.output.concentrate_ag  final.output.concentrate_pb  \
count                 16788.000000                 16788.000000   
mean                      4.716907                     9.113559   
std                       2.096718                     3.389495   
min                       0.000000                     0.000000   
25%                       3.971262                     8.825748   
50%                       4.869346                    10.065316   
75%                       5.821176                    11.054809   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  16490.000000                 16789.000000   
mean                       8.301123                    39.467217   
std                        3.825760                    13.917227   
min                        0.000000                     0.000000   
25%                        6.939185                    42.055722   
50%                        8.557228                    44.498874   
75%                       10.289741                    45.976222   
max                       18.124851                    53.611374   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           15339.000000          16794.000000          16677.000000   
mean               67.213166              8.757048              2.360327   
std                11.960446              3.634103              1.215576   
min                 0.000000              0.000000              0.000000   
25%                62.625685              7.610544              1.641604   
50%                67.644601              9.220393              2.453690   
75%                72.824595             10.971110              3.192404   
max               100.000000             19.552149              6.086532   

       final.output.tail_sol  final.output.tail_au  \
count           16715.000000          16794.000000   
mean                9.303932              2.687512   
std                 4.263208              1.272757   
min                 0.000000              0.000000   
25%                 7.870275              2.172953   
50%                10.021968              2.781132   
75%                11.648573              3.416936   
max                22.317730              9.789625   

       primary_cleaner.input.sulfate  ...  \
count                   15553.000000  ...   
mean                      129.479789  ...   
std                        45.386931  ...   
min                         0.000003  ...   
25%                       103.064021  ...   
50%                       131.783108  ...   
75%                       159.539839  ...   
max                       251.999948  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              16731.000000   
mean                                  19.101874   
std                                    6.883163   
min                                    0.000000   
25%                                   14.508299   
50%                                   19.986958   
75%                                   24.983961   
max                                   60.000000   

       secondary_cleaner.state.floatbank4_a_level  \
count                                16747.000000   
mean                                  -494.164481   
std                                     84.803334   
min                                   -799.920713   
25%                                   -500.837689   
50%                                   -499.778379   
75%                                   -494.648754   
max                                   -127.692333   

       secondary_cleaner.state.floatbank4_b_air  \
count                              16768.000000   
mean                                  14.778164   
std                                    5.999149   
min                                    0.000000   
25%                


Describe include: all :



date  final.output.concentrate_ag  \
count                 16860                 16788.000000   
unique                16860                          NaN   
top     2016-01-15 00:00:00                          NaN   
freq                      1                          NaN   
mean                    NaN                     4.716907   
std                     NaN                     2.096718   
min                     NaN                     0.000000   
25%                     NaN                     3.971262   
50%                     NaN                     4.869346   
75%                     NaN                     5.821176   
max                     NaN                    16.001945   

        final.output.concentrate_pb  final.output.concentrate_sol  \
count                  16788.000000                  16490.000000   
unique                          NaN                           NaN   
top                             NaN                           NaN   
freq                            NaN                           NaN   
mean                       9.113559                      8.301123   
std                        3.389495                      3.825760   
min                        0.000000                      0.000000   
25%                        8.825748                      6.939185   
50%                       10.065316                      8.557228   
75%                       11.054809                     10.289741   
max                       17.031899                     18.124851   

        final.output.concentrate_au  final.output.recovery  \
count                  16789.000000           15339.000000   
unique                          NaN                    NaN   
top                             NaN                    NaN   
freq                            NaN                    NaN   
mean                      39.467217              67.213166   
std                       13.917227              11.960446   
min                        0.000000               0.000000   
25%                       42.055722              62.625685   
50%                       44.498874              67.644601   
75%                       45.976222              72.824595   
max                       53.611374             100.000000   

        final.output.tail_ag  final.output.tail_pb  final.output.tail_sol  \
count           16794.000000          16677.000000           16715.000000   
unique                   NaN                   NaN                    NaN   
top                      NaN                   NaN                    NaN   
freq                     NaN                   NaN                    NaN   
mean                8.757048              2.360327               9.303932   
std                 3.634103              1.215576               4.263208   
min                 0.000000              0.000000               0.000000   
25%                 7.610544              1.641604               7.870275   
50%                 9.220393              2.453690              10.021968   
75%                10.971110              3.192404              11.648573   
max                19.552149              6.086532              22.317730   

        final.output.tail_au  ...  secondary_cleaner.state.floatbank4_a_air  \
count           16794.000000  ...                              16731.000000   
unique                   NaN  ...                                       NaN   
top                      NaN  ...                                       NaN   
freq                     NaN  ...                                       NaN   
mean                2.687512  ...                                 19.101874   
std                 1.272757  ...                                  6.883163   
min                 0.000000  ...                                  0.000000   
25%                 2.172953  ...                                 14.508299   
50%                 2.781132  ...                                 19.986958   
75%          


nulls in the columns:

final.output.concentrate_ag has 72 nulls, which is 0.4270% percent of Nulls
final.output.concentrate_pb has 72 nulls, which is 0.4270% percent of Nulls
final.output.concentrate_sol has 370 nulls, which is 2.1945% percent of Nulls
final.output.concentrate_au has 71 nulls, which is 0.4211% percent of Nulls
final.output.recovery has 1521 nulls, which is 9.0214% percent of Nulls
final.output.tail_ag has 66 nulls, which is 0.3915% percent of Nulls
final.output.tail_pb has 183 nulls, which is 1.0854% percent of Nulls
final.output.tail_sol has 145 nulls, which is 0.8600% percent of Nulls
final.output.tail_au has 66 nulls, which is 0.3915% percent of Nulls
primary_cleaner.input.sulfate has 1307 nulls, which is 7.7521% percent of Nulls
primary_cleaner.input.depressant has 1262 nulls, which is 7.4852% percent of Nulls
primary_cleaner.input.xanthate has 985 nulls, which is 5.8422% percent of Nulls
primary_cleaner.output.concentrate_ag has 82 nulls, which is 0.4864% percent

None


Zeros in the columns:

date 0
final.output.concentrate_ag 1263
final.output.concentrate_pb 1263
final.output.concentrate_sol 1263
final.output.concentrate_au 1263
final.output.recovery 89
final.output.tail_ag 1658
final.output.tail_pb 1658
final.output.tail_sol 1658
final.output.tail_au 1658
primary_cleaner.input.sulfate 0
primary_cleaner.input.depressant 57
primary_cleaner.input.feed_size 0
primary_cleaner.input.xanthate 0
primary_cleaner.output.concentrate_ag 1230
primary_cleaner.output.concentrate_pb 1230
primary_cleaner.output.concentrate_sol 1230
primary_cleaner.output.concentrate_au 1230
primary_cleaner.output.tail_ag 1549
primary_cleaner.output.tail_pb 1549
primary_cleaner.output.tail_sol 1549
primary_cleaner.output.tail_au 1549
primary_cleaner.state.floatbank8_a_air 256
primary_cleaner.state.floatbank8_a_level 0
primary_cleaner.state.floatbank8_b_air 257
primary_cleaner.state.floatbank8_b_level 0
primary_cleaner.state.floatbank8_c_air 282
primary_cleaner.state.floatbank8_c_lev

None


Shape: (16860, 87)


Duplicated: We have 0 duplicated rows


Dtypes:


date                                           object
final.output.concentrate_ag                   float64
final.output.concentrate_pb                   float64
final.output.concentrate_sol                  float64
final.output.concentrate_au                   float64
                                               ...   
secondary_cleaner.state.floatbank5_a_level    float64
secondary_cleaner.state.floatbank5_b_air      float64
secondary_cleaner.state.floatbank5_b_level    float64
secondary_cleaner.state.floatbank6_a_air      float64
secondary_cleaner.state.floatbank6_a_level    float64
Length: 87, dtype: object

notes
- There are 87 columns in the data
- All values are floats
- Some values are negative. They are reading of levels indicators. Since it's relative information it make sense
- There are 16860 observations in the data
- Almost all columns contains Non values
- Part of the columns contains some zeroes
- No duplicates rows

1.1.3.2  data_test

In [8]:
#print our info data
print('information about "data_test" dataset:')
general_info(data_test)

information about "data_test" dataset:



type  count   na       na%
date                                         object   5856    0  0.000000
primary_cleaner.input.sulfate               float64   5554  302  0.051571
primary_cleaner.input.depressant            float64   5572  284  0.048497
primary_cleaner.input.feed_size             float64   5856    0  0.000000
primary_cleaner.input.xanthate              float64   5690  166  0.028347
primary_cleaner.state.floatbank8_a_air      float64   5840   16  0.002732
primary_cleaner.state.floatbank8_a_level    float64   5840   16  0.002732
primary_cleaner.state.floatbank8_b_air      float64   5840   16  0.002732
primary_cleaner.state.floatbank8_b_level    float64   5840   16  0.002732
primary_cleaner.state.floatbank8_c_air      float64   5840   16  0.002732
primary_cleaner.state.floatbank8_c_level    float64   5840   16  0.002732
primary_cleaner.state.floatbank8_d_air      float64   5840   16  0.002732
primary_cleaner.state.floatbank8_d_level    float64   5840   16  0.002732
rougher.input.feed_ag                       float64   5840   16  0.002732
rougher.input.feed_pb                       float64   5840   16  0.002732
rougher.input.feed_rate                     float64   5816   40  0.006831
rougher.input.feed_size                     float64   5834   22  0.003757
rougher.input.feed_sol                      float64   5789   67  0.011441
rougher.input.feed_au                       float64   5840   16  0.002732
rougher.input.floatbank10_sulfate           float64   5599  257  0.043887
rougher.input.floatbank10_xanthate          float64   5733  123  0.021004
rougher.input.floatbank11_sulfate           float64   5801   55  0.009392
rougher.input.floatbank11_xanthate          float64   5503  353  0.060280
rougher.state.floatbank10_a_air             float64   5839   17  0.002903
rougher.state.floatbank10_a_level           float64   5840   16  0.002732
rougher.state.floatbank10_b_air             float64   5839   17  0.002903
rougher.state.floatbank10_b_level           float64   5840   16  0.002732
rougher.state.floatbank10_c_air             float64   5839   17  0.002903
rougher.state.floatbank10_c_level           float64   5840   16  0.002732
rougher.state.floatbank10_d_air             float64   5839   17  0.002903
rougher.state.floatbank10_d_level           float64   5840   16  0.002732
rougher.state.floatbank10_e_air             float64   5839   17  0.002903
rougher.state.floatbank10_e_level           float64   5840   16  0.002732
rougher.state.floatbank10_f_air             float64   5839   17  0.002903
rougher.state.floatbank10_f_level           float64   5840   16  0.002732
secondary_cleaner.state.floatbank2_a_air    float64   5836   20  0.003415
secondary_cleaner.state.floatbank2_a_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank2_b_air    float64   5833   23  0.003928
secondary_cleaner.state.floatbank2_b_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank3_a_air    float64   5822   34  0.005806
secondary_cleaner.state.floatbank3_a_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank3_b_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank3_b_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank4_a_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank4_a_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank4_b_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank4_b_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank5_a_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank5_a_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank5_b_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank5_b_level  float64   5840   16  0.002732
secondary_cleaner.state.floatbank6_a_air    float64   5840   16  0.002732
secondary_cleaner.state.floatbank6_a_level  float64   5840   16  0.002732


Head:


date  primary_cleaner.input.sulfate  \
0  2016-09-01 00:59:59                     210.800909   
1  2016-09-01 01:59:59                     215.392455   
2  2016-09-01 02:59:59                     215.259946   
3  2016-09-01 03:59:59                     215.336236   
4  2016-09-01 04:59:59                     199.099327   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                         14.993118                         8.080000   
1                         14.987471                         8.080000   
2                         12.884934                         7.786667   
3                         12.006805                         7.640000   
4                         10.682530                         7.530000   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        1.005021                             1398.981301   
1                        0.990469                             1398.777912   
2                        0.996043                             1398.493666   
3                        0.863514                             1399.618111   
4                        0.805575                             1401.268123   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.225577   
1                               -500.057435   
2                               -500.868360   
3                               -498.863574   
4                               -500.808305   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.144926   
1                             1398.055362   
2                             1398.860436   
3                             1397.440120   
4                             1398.128818   

   primary_cleaner.state.floatbank8_b_level  \
0                               -499.919735   
1                               -499.778182   
2                               -499.764529   
3                               -499.211024   
4                               -499.504543   

   primary_cleaner.state.floatbank8_c_air  ...  \
0                             1400.102998  ...   
1                             1396.151033  ...   
2                             1398.075709  ...   
3                             1400.129303  ...   
4                             1402.172226  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 12.023554   
1                                 12.058140   
2                                 11.962366   
3                                 12.033091   
4                                 12.025367   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -497.795834   
1                                 -498.695773   
2                                 -498.767484   
3                                 -498.350935   
4                                 -500.786497   

   secondary_cleaner.state.floatbank4_b_air  \
0                                  8.016656   
1                                  8.130979   
2                                  8.096893   
3                                  8.074946   
4                                  8.054678   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -501.289139   
1                                 -499.634209   
2                                 -500.827423   
3                                 -499.474407   
4                                 -500.397500   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  7.946562   
1                                  7.958270   
2                                  8.071056   
3                                  7.897085   
4                                  8.107890   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -432.317850   
1                                 -525.839648   
2                                 -500.801673   
3     


Tail:


date  primary_cleaner.input.sulfate  \
5851  2017-12-31 19:59:59                     173.957757   
5852  2017-12-31 20:59:59                     172.910270   
5853  2017-12-31 21:59:59                     171.135718   
5854  2017-12-31 22:59:59                     179.697158   
5855  2017-12-31 23:59:59                     181.556856   

      primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
5851                         15.963399                             8.07   
5852                         16.002605                             8.07   
5853                         15.993669                             8.07   
5854                         15.438979                             8.07   
5855                         14.995850                             8.07   

      primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
5851                        0.896701                             1401.930554   
5852                        0.896519                             1447.075722   
5853                        1.165996                             1498.836182   
5854                        1.501068                             1498.466243   
5855                        1.623454                             1498.096303   

      primary_cleaner.state.floatbank8_a_level  \
5851                               -499.728848   
5852                               -494.716823   
5853                               -501.770403   
5854                               -500.483984   
5855                               -499.796922   

      primary_cleaner.state.floatbank8_b_air  \
5851                             1401.441445   
5852                             1448.851892   
5853                             1499.572353   
5854                             1497.986986   
5855                             1501.743791   

      primary_cleaner.state.floatbank8_b_level  \
5851                               -499.193423   
5852                               -465.963026   
5853                               -495.516347   
5854                               -519.200340   
5855                               -505.146931   

      primary_cleaner.state.floatbank8_c_air  ...  \
5851                             1399.810313  ...   
5852                             1443.890424  ...   
5853                             1502.749213  ...   
5854                             1496.569047  ...   
5855                             1499.535978  ...   

      secondary_cleaner.state.floatbank4_a_air  \
5851                                 13.995957   
5852                                 16.749781   
5853                                 19.994130   
5854                                 19.958760   
5855                                 20.034715   

      secondary_cleaner.state.floatbank4_a_level  \
5851                                 -500.157454   
5852                                 -496.031539   
5853                                 -499.791312   
5854                                 -499.958750   
5855                                 -500.728588   

      secondary_cleaner.state.floatbank4_b_air  \
5851                                 12.069155   
5852                                 13.365371   
5853                                 15.101425   
5854                                 15.026853   
5855                                 14.914199   

      secondary_cleaner.state.floatbank4_b_level  \
5851                                 -499.673279   
5852                                 -499.122723   
5853                                 -499.936252   
5854                                 -499.723143   
5855                                 -499.948518   

      secondary_cleaner.state.floatbank5_a_air  \
5851                                  7.977259   
5852                                  9.288553   
5853                                 10.989181   
5854                                 11.011607   
5855                                 10.


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5856 entries, 0 to 5855
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        5856 non-null   object 
 1   primary_cleaner.input.sulfate               5554 non-null   float64
 2   primary_cleaner.input.depressant            5572 non-null   float64
 3   primary_cleaner.input.feed_size             5856 non-null   float64
 4   primary_cleaner.input.xanthate              5690 non-null   float64
 5   primary_cleaner.state.floatbank8_a_air      5840 non-null   float64
 6   primary_cleaner.state.floatbank8_a_level    5840 non-null   float64
 7   primary_cleaner.state.floatbank8_b_air      5840 non-null   float64
 8   primary_cleaner.state.floatbank8_b_level    5840 non-null   float64
 9   primary_cleaner.state.floatbank8_c_air      5840 non-null   float64
 10  prim

None


Describe:



primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                    5554.000000                       5572.000000   
mean                      170.515243                          8.482873   
std                        49.608602                          3.353105   
min                         0.000103                          0.000031   
25%                       143.340022                          6.411500   
50%                       176.103893                          8.023252   
75%                       207.240761                         10.017725   
max                       274.409626                         40.024582   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                      5856.000000                     5690.000000   
mean                          7.264651                        1.321420   
std                           0.611526                        0.693246   
min                           5.650000                        0.000003   
25%                           6.885625                        0.888769   
50%                           7.259333                        1.183362   
75%                           7.650000                        1.763797   
max                          15.500000                        5.433169   

       primary_cleaner.state.floatbank8_a_air  \
count                             5840.000000   
mean                              1481.990241   
std                                310.453166   
min                                  0.000000   
25%                               1497.190681   
50%                               1554.659783   
75%                               1601.681656   
max                               2212.432090   

       primary_cleaner.state.floatbank8_a_level  \
count                               5840.000000   
mean                                -509.057796   
std                                   61.339256   
min                                 -799.773788   
25%                                 -500.455211   
50%                                 -499.997402   
75%                                 -499.575313   
max                                  -57.195404   

       primary_cleaner.state.floatbank8_b_air  \
count                             5840.000000   
mean                              1486.908670   
std                                313.224286   
min                                  0.000000   
25%                               1497.150234   
50%                               1553.268084   
75%                               1601.784707   
max                               1975.147923   

       primary_cleaner.state.floatbank8_b_level  \
count                               5840.000000   
mean                                -511.743956   
std                                   67.139074   
min                                 -800.029078   
25%                                 -500.936639   
50%                                 -500.066588   
75%                                 -499.323361   
max                                 -142.527229   

       primary_cleaner.state.floatbank8_c_air  \
count                             5840.000000   
mean                              1468.495216   
std                                309.980748   
min                                  0.000000   
25%                               1437.050321   
50%                               1546.160672   
75%                               1600.785573   
max                               1715.053773   

       primary_cleaner.state.floatbank8_c_level  ...  \
count                               5840.000000  ...   
mean                                -509.741212  ...   
std                                   62.671873  ...   
min                                 -799.995127  ...   
25%                                 -501.300441  ...   
50%                                 -500.079537  ...   
75%                                


Describe include: all :



date  primary_cleaner.input.sulfate  \
count                  5856                    5554.000000   
unique                 5856                            NaN   
top     2016-09-01 00:59:59                            NaN   
freq                      1                            NaN   
mean                    NaN                     170.515243   
std                     NaN                      49.608602   
min                     NaN                       0.000103   
25%                     NaN                     143.340022   
50%                     NaN                     176.103893   
75%                     NaN                     207.240761   
max                     NaN                     274.409626   

        primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
count                        5572.000000                      5856.000000   
unique                               NaN                              NaN   
top                                  NaN                              NaN   
freq                                 NaN                              NaN   
mean                            8.482873                         7.264651   
std                             3.353105                         0.611526   
min                             0.000031                         5.650000   
25%                             6.411500                         6.885625   
50%                             8.023252                         7.259333   
75%                            10.017725                         7.650000   
max                            40.024582                        15.500000   

        primary_cleaner.input.xanthate  \
count                      5690.000000   
unique                             NaN   
top                                NaN   
freq                               NaN   
mean                          1.321420   
std                           0.693246   
min                           0.000003   
25%                           0.888769   
50%                           1.183362   
75%                           1.763797   
max                           5.433169   

        primary_cleaner.state.floatbank8_a_air  \
count                              5840.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                               1481.990241   
std                                 310.453166   
min                                   0.000000   
25%                                1497.190681   
50%                                1554.659783   
75%                                1601.681656   
max                                2212.432090   

        primary_cleaner.state.floatbank8_a_level  \
count                                5840.000000   
unique                                       NaN   
top                                          NaN   
freq                                         NaN   
mean                                 -509.057796   
std                                    61.339256   
min                                  -799.773788   
25%                                  -500.455211   
50%                                  -499.997402   
75%                                  -499.575313   
max                                   -57.195404   

        primary_cleaner.state.floatbank8_b_air  \
count                              5840.000000   
unique                                     NaN   
top                                        NaN   
freq                                       NaN   
mean                               1486.908670   
std                                 313.224286   
min                                   0.000000   
25%                                1497.150234   
50%                                1553.268084   
75%                                1601.784707   
max                                1975.147923   

        primary_clean


nulls in the columns:

primary_cleaner.input.sulfate has 302 nulls, which is 5.1571% percent of Nulls
primary_cleaner.input.depressant has 284 nulls, which is 4.8497% percent of Nulls
primary_cleaner.input.xanthate has 166 nulls, which is 2.8347% percent of Nulls
primary_cleaner.state.floatbank8_a_air has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_a_level has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_b_air has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_b_level has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_c_air has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_c_level has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_d_air has 16 nulls, which is 0.2732% percent of Nulls
primary_cleaner.state.floatbank8_d_level has 16 nulls, which is 0.2732% percent of Nulls
rougher.input.feed_ag has 16 nu

None


Zeros in the columns:

date 0
primary_cleaner.input.sulfate 0
primary_cleaner.input.depressant 0
primary_cleaner.input.feed_size 0
primary_cleaner.input.xanthate 0
primary_cleaner.state.floatbank8_a_air 93
primary_cleaner.state.floatbank8_a_level 0
primary_cleaner.state.floatbank8_b_air 88
primary_cleaner.state.floatbank8_b_level 0
primary_cleaner.state.floatbank8_c_air 48
primary_cleaner.state.floatbank8_c_level 0
primary_cleaner.state.floatbank8_d_air 48
primary_cleaner.state.floatbank8_d_level 0
rougher.input.feed_ag 369
rougher.input.feed_pb 369
rougher.input.feed_rate 0
rougher.input.feed_size 0
rougher.input.feed_sol 369
rougher.input.feed_au 369
rougher.input.floatbank10_sulfate 0
rougher.input.floatbank10_xanthate 0
rougher.input.floatbank11_sulfate 0
rougher.input.floatbank11_xanthate 0
rougher.state.floatbank10_a_air 0
rougher.state.floatbank10_a_level 0
rougher.state.floatbank10_b_air 0
rougher.state.floatbank10_b_level 0
rougher.state.floatbank10_c_air 0
rougher.state.floa

None


Shape: (5856, 53)


Duplicated: We have 0 duplicated rows


Dtypes:


date                                           object
primary_cleaner.input.sulfate                 float64
primary_cleaner.input.depressant              float64
primary_cleaner.input.feed_size               float64
primary_cleaner.input.xanthate                float64
primary_cleaner.state.floatbank8_a_air        float64
primary_cleaner.state.floatbank8_a_level      float64
primary_cleaner.state.floatbank8_b_air        float64
primary_cleaner.state.floatbank8_b_level      float64
primary_cleaner.state.floatbank8_c_air        float64
primary_cleaner.state.floatbank8_c_level      float64
primary_cleaner.state.floatbank8_d_air        float64
primary_cleaner.state.floatbank8_d_level      float64
rougher.input.feed_ag                         float64
rougher.input.feed_pb                         float64
rougher.input.feed_rate                       float64
rougher.input.feed_size                       float64
rougher.input.feed_sol                        float64
rougher.input.feed_au       

notes
- There are 53 columns in the data
- All values are floats
- Some values are negative. They are reading of levels indicators. Since it's relative information it make sense
- There are 5856 observations in the data
- Almost all columns contains Non values
- Part of the columns contains some zeroes
- No duplicates rows

## Check that recovery is calculated correctly. Using the training set, calculate recovery for the rougher.output.recovery feature. Find the MAE between your calculations and the feature values. Provide findings.


![recovery](img/recovery.jpg)

In this section I will:

- Create new column with my calculation for rougher.output.recovery
- Calculate MAE for my column (prediction) vs the original (target)
- For this section use only rows without Non in the relevant column:
rougher.output.concentrate_au , rougher.input.feed_au ,  rougher.output.tail_au , rougher.output.recovery


In [9]:
# drop nan in all relevant columns in the training set for this calculation 
data_train_remove_nan = data_train.dropna(
    subset=['rougher.output.concentrate_au',
    'rougher.input.feed_au',
    'rougher.output.tail_au',
    'rougher.output.recovery'])

In [10]:
data_train_remove_nan[['rougher.output.concentrate_au',
    'rougher.input.feed_au',
    'rougher.output.tail_au',
    'rougher.output.recovery']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14287 entries, 0 to 16859
Data columns (total 4 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   rougher.output.concentrate_au  14287 non-null  float64
 1   rougher.input.feed_au          14287 non-null  float64
 2   rougher.output.tail_au         14287 non-null  float64
 3   rougher.output.recovery        14287 non-null  float64
dtypes: float64(4)
memory usage: 558.1 KB


ok

In [11]:
C = 'rougher.output.concentrate_au'
T = 'rougher.output.tail_au'
F = 'rougher.input.feed_au'


data_train_remove_nan['new_rougher.output.recovery'] = ((
    data_train_remove_nan[C] * (
        data_train_remove_nan[F] - data_train_remove_nan[T]) ) \
/ (data_train_remove_nan[F] * (
    data_train_remove_nan[C] - data_train_remove_nan[T]) ))*100

In [12]:
mae = mean_absolute_error(y_true=data_train_remove_nan['rougher.output.recovery'],
                    y_pred=data_train_remove_nan['new_rougher.output.recovery'])

In [13]:
print('The MAE for the calculation of rougher recovery is: {:e}' .format(mae))

The MAE for the calculation of rougher recovery is: 9.303416e-15


The calculation predict exactly the recovery in the data. 

## Analyze the features not available in the test set. What are these parameters? What is their type?

In this section I will:
- Create a data frame using full data as a base with all features that are absent from the test.
- Check the info of this data frame
- Provide findings in text


In [ ]:
# list of missing columns in the data_test include the date

# connect all columns names
column_list_with_duplicates  = list(data_full.columns.values) + list(data_test.columns.values)
#drop duplicates and leve date
only_missing_columns_list = [
    x for x in column_list_with_duplicates if column_list_with_duplicates.count(x)==1 or x == 'date']
# remove last value - it is date . The first value is also date
only_missing_columns_list = only_missing_columns_list[:-1]

In [ ]:
# create data frame with only missing column
data_full_34_columns = data_full[only_missing_columns_list]

In [ ]:
# add the missing 34 columns to the data_test to have all the removed features back in the test set
data_test_with_missing = data_test.merge(right=data_full_34_columns, how='left' ,on='date')

# create df of test set with only the features that was missing
data_test_only_missing = data_test_with_missing[only_missing_columns_list]

In [ ]:
# print the type of the features that are not available in the test set
data_test_only_missing.info()

In [ ]:
# print the info on the data_test
data_test.info()

### Notes

- All the parameters are floats
- The removed features include all calculation features
- The test set contains only the inputs to the process and process state parameters - **not any output** so from the test set all process output was removed


- Since the features of the training and the set must match to use an ML model a decision of how to treat the missing column must be taken. I will use the training without the features that are missing in the test set. The model will be base on the process input and process state values only!


## Perform data preprocessing

### Retain the originals data frames with the addition of _raw to their name

In [ ]:
data_full_raw = data_full.copy(deep=True)
data_test_raw = data_test.copy(deep=True)
data_train_raw = data_train.copy(deep=True)

### EDA to determine how to treat the data

In [ ]:
# Check how many nan in this set
data_full_na = data_full.isna().sum().to_frame()

fig = px.bar(data_full_na, orientation='h', width=900, height=2000)
fig.show()

There are lots of missing values in the data. I will observe the reading of the features. The problem is that producing scatter plots  and histograms of all the observations is to heavy. The solution I choose is performing downsampling. I will leave only 1 out of every 100 observations and do it systematically while reading from the csv using the skiprows argument. I will create new data frames for the full csv this way.

In [ ]:
# a function that return true if the index modulue for 100 in the opend dataframe is not 0. 
# This way it will skip all rows beside the one their index modulue 100 is 0.
def skip(index):
    if index % 100 != 0:
       return True

In [ ]:
try:
    data_full_downsample = pd.read_csv('gold_recovery_full.csv', skiprows= lambda x: skip(x))
except:
    data_full_downsample = pd.read_csv('/datasets/gold_recovery_full.csv', skiprows= lambda x: skip(x))
    
print('Data has been read correctly!')

In [ ]:
# produce scatter for one of ~10 of the downsample features
skipping_interval = int(data_full_downsample.shape[1]/10)
counter = 1
for column in data_full_downsample.columns:
    if counter % skipping_interval == 0:
        print(f'{column} scatter and histogram')
        px.scatter(data_full_downsample, x='date', y=column).show()
        px.histogram(data_full_downsample, x=column).show()
        print()
        print()
    
    counter += 1

- When observing the data we can see that the values are generally represent continues process and therefore if we want to replace nan we better do it base on interpolate values. 
- There are some extreme values. My guess is that this is a noise that comes from the analysers. I will leave them since this represents the process and in any given test set and future predictions they will appear and I will have to let the model figure out how to treat them.
- Values are in different scale. This will interfere any regression model. Therefore I will scale my features.

### Add target columns to the test set from the full columns base on indexes

In [ ]:
# columns to add and date that is the ref column
recovery_strings = ['date', 'final.output.recovery', 'rougher.output.recovery']

In [ ]:
data_test = data_test.merge(right=data_full[recovery_strings], 
                           how='left',
                           on='date')

### Handling Missing Data

In this part I will remove all observation with no target because we can't reliably fill them following that I will forward fill all the other missing values 

In [ ]:
# remove all observation with no target
data_full.dropna(subset=['final.output.recovery', 'rougher.output.recovery'], inplace=True)
data_test.dropna(subset=['final.output.recovery', 'rougher.output.recovery'], inplace=True)
data_train.dropna(subset=['final.output.recovery', 'rougher.output.recovery'], inplace=True)

In [ ]:
# Remove zero recovery. This is not valid value for the smape calculation
data_full = data_full[
    (data_full['final.output.recovery'] != 0) & (data_full['rougher.output.recovery'] != 0)
]

data_test = data_test[
    (data_test['final.output.recovery'] != 0) & (data_test['rougher.output.recovery'] != 0)
]

data_train = data_train[
    (data_train['final.output.recovery'] != 0) & (data_train['rougher.output.recovery'] != 0)
]


In [ ]:
# forward fill rest of the columns
data_full.fillna(method='ffill', inplace=True)
data_train.fillna(method='ffill', inplace=True)
data_test.fillna(method='ffill', inplace=True)

### Add  datetime column 

In [ ]:
data_train['date_datetime'] = pd.to_datetime(data_train['date'], format='%Y-%m-%d %H:%M:%S') 
data_test['date_datetime'] = pd.to_datetime(data_test['date'], format='%Y-%m-%d %H:%M:%S') 
data_full['date_datetime'] = pd.to_datetime(data_test['date'], format='%Y-%m-%d %H:%M:%S') 

# Analyze the data

## Take note of how the concentrations of metals (Au, Ag, Pb) change depending on the purification stage


In this section I will do the following:
- Plot histogram of the metals concentration with line for each purification stage 
- Explain the graph

In [ ]:
# set a grey background (use sns.set_theme() if seaborn version 0.11.0 or above) 
sns.set(style="darkgrid", rc={'figure.figsize':(14, 6)})

sns.histplot(
    data=data_full, x="rougher.input.feed_ag",
    color='r', label="rougher.input.feed_ag", kde=True);
sns.histplot(
    data=data_full, x="rougher.output.concentrate_ag",
    color='g', label="rougher.output.concentrate_ag", kde=True);
sns.histplot(
    data=data_full, x="primary_cleaner.output.concentrate_ag",
    color='b', label="primary_cleaner.output.concentrate_ag", kde=True);
ax=sns.histplot(
    data=data_full, x="final.output.concentrate_ag",
    color='y', label="final.output.concentrate_ag", kde=True);

plt.legend();
ax.set_title('Ag concentrations',fontsize=20)
ax.set_xlabel("concentrations",fontsize=15)
ax.set_ylabel("counts",fontsize=15)
# aa.set(xlabel='common xlabel', ylabel='common ylabel')
plt.show();


- The ag concentration increases in the flotation stage. 
- In the two later stage the concentration decreased a little bit. 
- It is clear the process does not increase the concentration of ag in the process line as should be expected since this is not a process to extract ag.

In [ ]:
# set a grey background (use sns.set_theme() if seaborn version 0.11.0 or above) 
sns.set(style="darkgrid", rc={'figure.figsize':(14, 6)})

sns.histplot(
    data=data_full, x="rougher.input.feed_pb",
    color='r', label="rougher.input.feed_pb", kde=True);
sns.histplot(
    data=data_full, x="rougher.output.concentrate_pb",
    color='g', label="rougher.output.concentrate_pb", kde=True);
sns.histplot(
    data=data_full, x="primary_cleaner.output.concentrate_pb",
    color='b', label="primary_cleaner.output.concentrate_pb", kde=True);
ax=sns.histplot(
    data=data_full, x="final.output.concentrate_pb",
    color='y', label="final.output.concentrate_pb", kde=True);

plt.legend();
ax.set_title('Pb concentrations',fontsize=20)
ax.set_xlabel("concentrations",fontsize=15)
ax.set_ylabel("counts",fontsize=15)
# aa.set(xlabel='common xlabel', ylabel='common ylabel')
plt.show();

- pb concentration increased in the flotation from ~3 to ~8
- After that in the last two stages it increased a tiny bit to ~10
- This is because this process is not focus on pb

In [ ]:
# set a grey background (use sns.set_theme() if seaborn version 0.11.0 or above) 
sns.set(style="darkgrid", rc={'figure.figsize':(14, 6)})

sns.histplot(
    data=data_full, x="rougher.input.feed_au",
    color='r', label="rougher.input.feed_au", kde=True);
sns.histplot(
    data=data_full, x="rougher.output.concentrate_au",
    color='g', label="rougher.output.concentrate_au", kde=True);
sns.histplot(
    data=data_full, x="primary_cleaner.output.concentrate_au",
    color='b', label="primary_cleaner.output.concentrate_au", kde=True);
ax=sns.histplot(
    data=data_full, x="final.output.concentrate_au",
    color='y', label="final.output.concentrate_au", kde=True);

plt.legend();
ax.set_title('Au concentrations',fontsize=20)
ax.set_xlabel("concentrations",fontsize=15)
ax.set_ylabel("counts",fontsize=15)
# aa.set(xlabel='common xlabel', ylabel='common ylabel')
plt.show();

- The au (gold) concentration increases in **all** stages. 
- The concentration increased from ~7 to ~20 in flotation stage. 
- In the primary cleaner the concentration increased to ~34.
- In the last stage the concentration increased to ~45!

### concentrations of metals conclusion
It is clear that the process is selective to gold. The other metals concentration almost don't change and for some of the cases it decreased. This show that for purifying the gold from the ore the other metals removed.


## Compare the feed particle size distributions in the training set and in the test set. If the distributions vary significantly, the model evaluation will be incorrect.


In this section I will do the following:

- Produce histograms of the particle size distributions in the training set and in the test set.
- Run t-test 
- If distributions are similar, it is safe to use the test to evaluate the trained model


In [ ]:
# Get the columns of size distribution
size_columns = [
    s for s in data_test.columns.to_list() if 'size' in s]
size_columns

- I will take only the rougher input because this is not affected by the process and our goal is only to determine if something out of the process happen different in times it was taken for test or for training so that there is different

In [ ]:
# set a grey background (use sns.set_theme() if seaborn version 0.11.0 or above) 
sns.set(style="darkgrid", rc={'figure.figsize':(14, 6)})

sns.histplot(
    data=data_test, x="rougher.input.feed_size",
    color='r', label="data_test_input.feed_size", kde=True);

sns.histplot(
    data=data_train, x="rougher.input.feed_size",
    color='b', label="data_train_input.feed_size", kde=True);

plt.legend();
ax.set_title('feed_size for test and train set',fontsize=20)
ax.set_xlabel("feed_size",fontsize=15)
ax.set_ylabel("counts",fontsize=15)
plt.show();

The particle size are similar, it is safe to use the test to evaluate the trained model

## Consider the total concentrations of all substances at different stages: raw feed, rougher concentrate, and final concentrate. Do you notice any abnormal values in the total distribution? If you do, is it worth removing such values from both samples? Describe the findings and eliminate anomalies.


To solve this one I will do the following

- All substances will be Au, Ag, Pb. 
- Combine their concentration in each stage to produce a new column called:       [stage].total_concentration
- Produce histogram
- Look for abnormalities such as:
 - In later stages the amount increases

In [ ]:
# create column  for each of the process stage
rougher_input_feed = [
 'rougher.input.feed_ag',
 'rougher.input.feed_pb',
 'rougher.input.feed_au'
]

rougher_output_concentrate = [
 'rougher.output.concentrate_ag',
 'rougher.output.concentrate_pb',
 'rougher.output.concentrate_au'
]

primary_cleaner_output_concentrate = [
 'primary_cleaner.output.concentrate_ag',
 'primary_cleaner.output.concentrate_pb',
 'primary_cleaner.output.concentrate_au'
]

final_output_concentrate = [
 'final.output.concentrate_ag',
 'final.output.concentrate_pb',
 'final.output.concentrate_au'
]

column_list = [
    'rougher_input_feed',
    'rougher_output_concentrate',
    'primary_cleaner_output_concentrate',
    'final_output_concentrate'
]

In [ ]:
# create df for only substance
df_for_ag_au_pb_total = data_full[[
 'final.output.concentrate_ag',
 'final.output.concentrate_pb',
 'final.output.concentrate_au',
 'primary_cleaner.output.concentrate_ag',
 'primary_cleaner.output.concentrate_pb',
 'primary_cleaner.output.concentrate_au',
 'rougher.input.feed_ag',
 'rougher.input.feed_pb',
 'rougher.input.feed_au',
 'rougher.output.concentrate_ag',
 'rougher.output.concentrate_pb',
 'rougher.output.concentrate_au'
]]

In [ ]:
# sum concentrations by stage
df_for_ag_au_pb_total['rougher_input_feed'] = df_for_ag_au_pb_total[
    rougher_input_feed].sum(axis=1) 
df_for_ag_au_pb_total['rougher_output_concentrate']= df_for_ag_au_pb_total[
    rougher_output_concentrate].sum(axis=1)
df_for_ag_au_pb_total['primary_cleaner_output_concentrate']= df_for_ag_au_pb_total[
    primary_cleaner_output_concentrate].sum(axis=1)
df_for_ag_au_pb_total['final_output_concentrate']= df_for_ag_au_pb_total[
    final_output_concentrate].sum(axis=1)

In [ ]:
# left only sum
df_for_ag_au_pb_total = df_for_ag_au_pb_total[column_list]
df_for_ag_au_pb_total

In [ ]:
# plot histograms

# Group data together
hist_data = [df_for_ag_au_pb_total['rougher_input_feed'],
             df_for_ag_au_pb_total['rougher_output_concentrate'],
             df_for_ag_au_pb_total['primary_cleaner_output_concentrate'],
             df_for_ag_au_pb_total['final_output_concentrate']]
        

group_labels = ['rougher_input_feed',
                'rougher_output_concentrate',
                'primary_cleaner_output_concentrate',
                'final_output_concentrate']

# Create distplot with custom bin_size
fig = ff.create_distplot(hist_data, group_labels, bin_size=.2)
fig.show()

the findings are:
- The total concentration increase from the feed to the output of the primary cleaner and is the highest in the final output
- The abnormality is that there are lots of zero values. I can see that they appear in all of the stages and there is no continuity between the main concentrations and the zero values thus we can say that these results cause by a measurement error from some kind. Since the concentration don't make it to our model we would not treat it but otherwise we should have remove them somehow. 


# Build the model

## Write a function to calculate the final sMAPE value.

What I will do here is:
- Write a function to calculate the sMAPE for the rougher and the final stage using the predicted recovery and target recovery
- test that the function works using the test set

In [ ]:
def smape(y_true, y_pred):
    return (np.abs(y_true-y_pred)/((np.abs(y_true)+np.abs(y_pred))/2)).mean()

First we will create features and target sets for the training set and for the test set 

In [ ]:
features_train = data_train[data_test.columns]
features_train = features_train.drop(['final.output.recovery',
                                      'rougher.output.recovery',
                                     'date',
                                     'date_datetime'], axis=1).values
target_train = data_train[['rougher.output.recovery', 'final.output.recovery']].values

In [ ]:
features_train.shape

In [ ]:
target_train.shape

In [ ]:
features_test = data_test.drop(['final.output.recovery',
                                      'rougher.output.recovery',
                                     'date',
                                     'date_datetime'], axis=1).values
target_test = data_test[['rougher.output.recovery', 'final.output.recovery']].values

In [ ]:
target_test.shape

In [ ]:
target_test.shape

In [ ]:
lr = LinearRegression()

In [ ]:
lr.fit(features_train, target_train)

In [ ]:
pred_lr = lr.predict(features_test)

In [ ]:
pred_lr[:, 0].shape

In [ ]:
smape(target_test[:,0], pred_lr[:,0])

In [ ]:
smape(target_test[:,1], pred_lr[:,1])

In [ ]:
def smape_final(y_true,y_pred):
    smape_out_rougher = smape(y_true[:,0], y_pred[:,0])
    smape_out_final = smape(y_true[:,1], y_pred[:,1])
    return 0.25*smape_out_rougher + 0.75*smape_out_final

In [ ]:
smape_final(target_test, pred_lr)

## Train different models. Evaluate them using cross-validation. Pick the best model and test it using the test sample. Provide findings

In [ ]:
# define scorer
smape_score = make_scorer(smape_final, greater_is_better=False)

Do GridSearchCV for LinearRegression

In [ ]:
lr.get_params()

In [ ]:
parameters = {'fit_intercept': [True, False]
             }

In [ ]:
lr = LinearRegression()

In [ ]:
# function to do GridSearchCV 
def gs_evaluate(model, params):
    gs = GridSearchCV(model, param_grid=params , cv=5, scoring=smape_score, verbose=0, refit=True)
    gs.fit(features_train, target_train)
    best_score = -gs.best_score_
    score_train = -gs.score(features_train, target_train)
    score_test = -gs.score(features_test, target_test)
    best_params = gs.best_params_
    return best_score, score_train, score_test, best_params

In [ ]:
best_score_lr, score_train_lr, score_test_lr, best_params_lr = gs_evaluate(lr, parameters)

In [ ]:
print('For Linear Regression the best score in the cross validation is {:0.3f}, \n\
the best score on all training set is {:0.3f}, \n\
the best score on the test set is {:0.3f} \n\
and the parameters obtained from the GridSearchCV are: \n{}'.format(
best_score_lr, score_train_lr, score_test_lr, best_params_lr
))

Do GridSearchCV for DecisionTreeRegressor

In [ ]:
dtr = DecisionTreeRegressor()

In [ ]:
dtr.get_params()

In [ ]:
parameters = {'max_depth': range(1,4),
          'min_impurity_decrease' : np.arange(.01, .20, .05),
          'min_samples_split' : range(2, 6, 2)}

In [ ]:
dtr = DecisionTreeRegressor(random_state=12345)

In [ ]:
best_score_dtr, score_train_dtr, score_test_dtr, best_params_dtr = gs_evaluate(dtr, parameters)

In [ ]:
print('For Decision Tree Regressor the best score in the cross validation is {:0.3f}, \n\
the best score on all training set is {:0.3f}, \n\
the best score on the test set is {:0.3f} \n\
and the parameters obtained from the GridSearchCV are: \n{}'.format(
best_score_dtr, score_train_dtr, score_test_dtr, best_params_dtr
))

Strange that the GridSearchCV choose 'max_depth': 1. Yet the results is a little better then of linear regression.

In [ ]:
rfr = RandomForestRegressor(random_state=12345)

In [ ]:
rfr.get_params()

In [ ]:
parameters = {'n_estimators' : range(20, 41, 10),
             'max_depth' : range(2, 5, 1)}

In [ ]:
best_score_rfr, score_train_rfr, score_test_rfr, best_params_rfr = gs_evaluate(rfr, parameters)

In [ ]:
print('For Random Forest Regressor the best score in the cross validation is {:0.3f}, \n\
the best score on all training set is {:0.3f}, \n\
the best score on the test set is {:0.3f} \n\
and the parameters obtained from the GridSearchCV are: \n{}'.format(
best_score_rfr, score_train_rfr, score_test_rfr, best_params_rfr
))

comparison with a simple baseline 

In [ ]:
dummy_regr = DummyRegressor(strategy="mean")
dummy_regr.fit(features_train, target_train)
dummy_pred = dummy_regr.predict(features_test)

In [ ]:
smpe_dummy = smape_final(target_test, dummy_pred)

In [ ]:
print('The dummy sMAPE is {:0.3f}'.format(smpe_dummy))

## Conclusion 
In this section after optimizing 3 models the best one is Random Forest Regressor. It gave same sMAPE score for the test set as the Decision Tree Regressor but for the training it worked slightly better. Meanwhile the linear regression showed results almost as those of the dummy regression. We have to admit that even our top results was not much better then the dummy test. Probably in this case the mean value of the target resemble quite good it predicted value since the chemical separation process is pretty stable.